api/traces/v1/http.go

```diff
@@ -163,6 +165,55 @@ func NewV2Handler(read *url.URL, opts ...HandlerOption) http.Handler {
 	return r
 }
 
+// Parse a URL; if the URL includes `{tenant}` that portion will be replaced by the tenant
+func ExpandTemplatedUpstream(templateUpstream, tenant string) (*url.URL, error) {
+	rawTracesReadEndpoint := strings.Replace(templateUpstream, "{tenant}", tenant, 1)
+	return url.ParseRequestURI(rawTracesReadEndpoint)
+}
+
+// middlewareSetTemplatedUpstream is a variation of proxy.MiddlewareSetUpstream()
+// with additional processing if the upstream includes "{tenant}"
+func middlewareSetTemplatedUpstream(logger log.Logger, upstreamTemplate string) proxy.Middleware {
```

Comment:
I'm a bit worried about the future of this code. The hope is that eventually Jaeger will natively support tenancy and we will be able to stop using templated strings in the `--endpoint` flag but this code will still live here, providing a kind of secret functionality without maintenance. I wonder if this is a special piece of Jaeger-specific stop-gap business logic that should be handled somewhere else like (sigh) a special-purpose proxy that we can later throw away. Actually, you could probably even configure nginx to do this transformation with minimal config.

Reply:
This PR is a temporal workaround because Jaeger does not support multitenancy natively. The plan is to add multitenancy to Jaeger, we are already working on it. 

We could slightly change this PR to add a new flag that would allow tenant template in the endpoint. It will clearly allow us to enable/disable the feature and deprecate once multitenancy is in place.

Reply:
Yes, I'm aware that this is temporary and that's exactly what makes me so worried. I'm worried that this temporary code will secretly live here forever because it's a forgotten hack.

I agree that adding an explicitly UX-fiendish flag that is tied to this hack would help isolate the code and improve the likelihood of getting this it removed once the feature is available in Jaeger. An example could be `--experimental.traces.read.endpoint-template`

Reply:
@squat I have moved the template stuff to an experimental flag as recommended by you and @pavolloffay .

Your point about using a proxy to keep this code out of Observatorium-api has merit.  I tried to start the discussion in https://github.com/observatorium/api/issues/268 .  Architecturally I agree with you.  I did it here because my feeling is that creating Jsonnet to configure a reverse proxy creates technical debt than the Golang to implement the template.

Reply:
Yea, I see both points, but to the Ed point I think it make sense to keep in in API if there is no need to reuse it in Jaeger. Let's make sure it is separate package, documented accordingly. Seems a bit better than jsonnet + another proxy yes.